Demo/MozillaSocial-iOS/MoSoContent/Auth.swift

```diff
@@ -0,0 +1,138 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import AuthenticationServices
+
+let isProd = false
+let baseURL = isProd ? "https://mozilla.social" : "https://stage.moztodon.nonprod.webservices.mozgcp.net"
+let redirectScheme = "mozillasocial"
+let redirectURI = redirectScheme + "://auth"
+
+class Auth: NSObject, ASWebAuthenticationPresentationContextProviding {
+
+    var authToken: Token?
+
+    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
+        return ASPresentationAnchor()
+    }
+
+    func registerApp() async -> ClientEntity? {
+        var urlRequest = URLRequest(url: registerAppURL())
+        urlRequest.httpMethod = "POST"
+
+        do{
+            let (data, _) = try await URLSession.shared.data(for: urlRequest)
+
+            let decoder = JSONDecoder()
+            decoder.keyDecodingStrategy = .convertFromSnakeCase
+            return try decoder.decode(ClientEntity.self, from: data)
+        } catch {
+            print("error: ", error)
+            return nil
+        }
+    }
+
+    func signIn(client: ClientEntity) {
+        var authURL = signInURL(with: client.clientId)
+
+        let session = ASWebAuthenticationSession(url: authURL, callbackURLScheme: redirectScheme) { (url, error) in
+            guard error == nil else {
+                print("ERROR: \(String(describing: error))")
+                return
+            }
+
+            guard let url = url, let code = self.parseCode(from: url) else {
+                print("Invalid Code Return")
+                return
+            }
+
+            Task {
+                self.authToken = await self.fetchAuthToken(for: client, with: code)
+            }
+        }
+
+        session.presentationContextProvider = self
+        session.prefersEphemeralWebBrowserSession = true
+        session.start()
+    }
+
+    private func fetchAuthToken(for client: ClientEntity, with code: String) async -> Token? {
```

Comment:
any chance this could become a throwing method, and we catch the error at the caller site? Probably doesn't matter much in this scenario, but Ideally a caller would know the error and decide how to handle it

Reply:
Same as previous: Happy Path!